cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -170,6 +185,106 @@ class OperandNode extends Node, TOperandNode {
   override string toString() { result = this.getOperand().toString() }
 }
 
+abstract private class SkippableInstruction extends Instruction {
+  abstract Instruction getSourceInstruction();
+}
+
+private Instruction skipSkippableInstructionsRec(SkippableInstruction skip) {
+  result = skip.getSourceInstruction() and not result instanceof SkippableInstruction
+  or
+  result = skipSkippableInstructionsRec(skip.getSourceInstruction())
+}
+
+private Instruction skipSkippableInstructions(Instruction instr) {
+  result = instr and not result instanceof SkippableInstruction
+  or
+  result = skipSkippableInstructionsRec(instr)
+}
+
+private class SkippableCopyValueInstruction extends SkippableInstruction, CopyValueInstruction {
+  override Instruction getSourceInstruction() { result = this.getSourceValue() }
+}
+
+private class SkippableConvertInstruction extends SkippableInstruction, ConvertInstruction {
+  override Instruction getSourceInstruction() { result = this.getUnary() }
+}
+
+private class SkippableCheckedConvertInstruction extends SkippableInstruction,
+  CheckedConvertOrNullInstruction {
+  override Instruction getSourceInstruction() { result = this.getUnary() }
+}
+
+private class SkippableInheritanceConversionInstruction extends SkippableInstruction,
+  InheritanceConversionInstruction {
+  override Instruction getSourceInstruction() { result = this.getUnary() }
+}
+
+/**
+ * INTERNAL: do not use. Gets the `FieldNode` corresponding to `instr`, if
+ * `instr` is an instruction that propagates an address of a `FieldAddressInstruction`.
+ */
+FieldNode getFieldNodeForFieldInstruction(Instruction instr) {
+  result.getFieldInstruction() = skipSkippableInstructions(instr)
+}
+
+/**
+ * INTERNAL: do not use. A `FieldNode` represents the state of an object after modifying one
+ * of its fields.
+ */
+class FieldNode extends Node, TFieldNode {
+  FieldAddressInstruction field;
+
+  FieldNode() { this = TFieldNode(field) }
+
+  /** Gets the `Field` of this `FieldNode`. */
+  Field getField() { result = getFieldInstruction().getField() }
+
+  /** Gets the `FieldAddressInstruction` of this `FieldNode`. */
+  FieldAddressInstruction getFieldInstruction() { result = field }
+
+  /**
+   * Gets the `FieldNode` corresponding to the parent field of this `FieldNode`, if any.
+   *
+   * For example, if `f` is the `FieldNode` for `c` in the expression `a.b.c`, then `f.getObjectNode()`
+   * gives the `FieldNode` of `b`, and `f.getObjectNode().getObjectNode()` has no result as `a` is
+   * not a field.
+   */
+  FieldNode getObjectNode() { result = getFieldNodeForFieldInstruction(field.getObjectAddress()) }
```

Comment:
The skippable instruction concept is really a bad version of a hypothetical `getUnqualifiedInstruction` that would get the definition of an operand and skip past all of the conversions.
It's probably fine to replace this concept with local data flow (or at least the union of `simpleInstructionLocalFlowStep` and `simpleOperandLocalFlowStep` - I'm not exactly sure what happens if we have flow in and out of `FieldNode`s when searching back in the dataflow graph for the instruction that defines the qualifier).

Reply:
I don't follow. Dereferences or loads that slice a field out of a struct won't be exact loads of the field address. Maybe they'll be loads of the field value, but that's something else.

What I mean is that
```c
a.b.c = source();
```
should be the same as
```c
B *b = &a.b;
b->c = source();
```
and that maybe that can be achieved if the `CopyValueInstruction` case is replaced with `CopyInstruction`. I haven't thought it through, though. The problem probably has to do with time travel: if we allow reverse reads to jump back in time across variable assignments, then too many things might have happened in between.

I'll keep it on a list of things to discuss in our next Zoom call.

Reply:
> I don't follow. Dereferences or loads that slice a field out of a struct won't be exact loads of the field address. Maybe they'll be loads of the field value, but that's something else.

Sorry, maybe I was unclear about where the problem was. The line`sink(a->f);` will have two loads: one will be an inexact load of `f` which we will register as a readStep (and pop `f` from the access path), and the other will be an exact load of `a` that lods from `InitializeIndirection[a]`. By skipping past that exact load we won't register the load of `a` as a read step that reads an `ArrayContent`.

But, yes. Let's discuss this over Zoom in our next meeting.

> and that maybe that can be achieved if the `CopyValueInstruction` case is replaced with `CopyInstruction`. I haven't thought it through, though. The problem probably has to do with time travel: if we allow reverse reads to jump back in time across variable assignments, then too many things might have happened in between.

You're right. That's exactly what will happen. One problem is that of time-travel, and another problem is that this relies on the IR alias analysis, and since this is a 'must'-alias it won't give us flow in this case:
```cpp
B *b = nullptr;
if(...) {
  b = &a.b;
}
b->c = source();
```

Reply:
Having had a couple of Christmas days to think this over, I've replaced the skippable instruction concept with local dataflow steps in 258d04178f90b48153c1119aa55a86544cbea54c. This eliminates the need for another PR I was planning on doing to address the snippet in https://github.com/github/codeql-c-analysis-team/issues/64#issuecomment-749472056.